oscilloscope/wrappers.py

```diff
@@ -0,0 +1,318 @@
+from __future__ import absolute_import
+
+from labrad import types as T
+from labrad.gpib import GPIBDeviceWrapper
+from twisted.internet.defer import inlineCallbacks, returnValue
+
+import numpy
+
+
+def require_is_in(obj, collection):
+    if obj not in collection:
+        raise ValueError("%s not in %s"%(obj, collection))
+
+
+class NotImplementedAttribute(object):
+    """An attribute which has not been implemented."""
+
+    def __init__(self, label):
+        self.label = label
+
+    def __get__(self, inst, cls):
+        """Raise AttributeError because this attribute should be overridden"""
+        raise AttributeError("Attribute {} of class {} not implemented".format(
+            self.label,
+            cls))
+
+
+class OscilloscopeWrapper(GPIBDeviceWrapper):
+    """Base class for oscilloscope wrappers
+
+    We provide iterpolatable strings which can be treated as functions to
+    generate VISA commands. For example, consider the string
+        'CH{0}:COUP {1}'.
+    This string has a .format method, which is a function that takes two
+    arguments. When you call that function with arguments 0 and "ON", the
+    result is
+        'CH0:COUP ON'
+    which is the VISA command that turns on channel 0.
+    """
+
+    # CHANNEL
+
+    # input
+
+    CHANNEL_STATES = ['ON', 'OFF']
+    channel_on_off_write = NotImplementedAttribute('channel_on_off_write')
+    channel_on_off_query = NotImplementedAttribute('channel_on_off_query')
+    channel_on_off_parse = NotImplementedAttribute('channel_on_off_parse')
+
+    COUPLINGS = ['AC', 'DC']
+    coupling_write = 'CH{0}:COUP {1}'.format
+    coupling_query = 'CH{0}:COUP?'.format
```

Comment:
The reason was a failed attempt to make some of the strings work on both Tek and Agilent devices.